File: jupyter2/problems/recursive_type/tttree/go/tttree.go
package main
import (
	"fmt"
	"slices"
	"strings"
)

/** begin my answer */

/** end my answer */

// add all values in list xs to tree t in order
func addSeq(xs []int, t *TTTree) *TTTree {
	for _, x := range xs {
		t = tttreeAdd(x, t)
	}
	return t
}

func lookupSeq(xs []int, t *TTTree) bool {
	for _, x := range xs {
		if !tttreeLookup(x, t) {
			return false
		}
	}
	return true
}

func removeSeq(xs []int, t *TTTree) *TTTree {
	for _, x := range xs {
		t = tttreeRemove(x, t)
	}
	return t
}

// 0 ... n - 1
func mkRange(n int) []int {
	s := make([]int, n)
	for i := 0; i < n; i++ {
		s[i] = i
	}
	return s
}

// random sequence
func randomSeq(n int, x int) []int {
	seq := make([]int, n)
	for i := 0; i < n; i++ {
		x = (0x5DEECE66D * x + 0xB) & 0xFFFFFFFFFFFF
		seq[i] = x >> 17
	}
	return seq
}

// randomly shuffle xs
type Pair struct {
	k int
	x int
}

func printSlice(header string, xs []int, n int) {
	fmt.Print(header, "[")
	for i, x := range xs {
		if i >= n {
			fmt.Print(" ...")
			break
		}
		if i > 0 {
			fmt.Print(", ")
		}
		fmt.Print(x)
	}
	fmt.Println("]")
}

func randomShuffle(seed int, xs []int) []int {
	ys := randomSeq(len(xs), seed)
	ps := make([]Pair, len(xs))
	for i, x := range xs {
		ps[i] = Pair{k: ys[i], x: x}
	}
	slices.SortFunc(ps, func(a, b Pair) int {
		return a.k - b.k
	})
	for i, z := range ps {
		ys[i] = z.x
	}
	return ys
}

func checkSeq(xs0 []int, xs1 []int) bool {
	printSlice("insert in this order: ", xs0, 7)
	printSlice("remove in this order: ", xs1, 7)
	t := addSeq(xs0, nil)
	tttreeCheck(t)
	if lookupSeq(xs1, t) {
		e := removeSeq(xs1, t)
		return e == nil
	} else {
		return false
	}
}

// randomly shuffle 0 .. n-1;
//   add them to empty tree;
//   remove all elements from the tree;
//   check they are sorted

func checkRandom(insertSeed int, removeSeed, n int) bool {
	rng := mkRange(n)
	xs0 := randomShuffle(insertSeed, rng)
	xs1 := randomShuffle(removeSeed, rng)
	return checkSeq(xs0, xs1)
}

func main() {
	insertSeed := 1234
	removeSeed := 123456
	if true {
		if !checkRandom(insertSeed, removeSeed, 2) { panic("wrong") }
		if !checkRandom(insertSeed, removeSeed, 3) { panic("wrong") }
		if !checkRandom(insertSeed + 0, removeSeed + 0, 4) { panic("wrong") }
		if !checkRandom(insertSeed + 1, removeSeed + 1, 4) { panic("wrong") }
		if !checkRandom(insertSeed + 2, removeSeed + 2, 4) { panic("wrong") }
		if !checkRandom(insertSeed + 3, removeSeed + 3, 4) { panic("wrong") }
		if !checkRandom(insertSeed + 0, removeSeed + 0, 5) { panic("wrong") }
		if !checkRandom(insertSeed + 1, removeSeed + 1, 5) { panic("wrong") }
		if !checkRandom(insertSeed + 2, removeSeed + 2, 5) { panic("wrong") }
		if !checkRandom(insertSeed + 3, removeSeed + 3, 5) { panic("wrong") }
		if !checkRandom(insertSeed, removeSeed, 10) { panic("wrong") }
	}
	if !checkRandom(insertSeed, removeSeed, 12) { panic("wrong") }
	if true {
		if !checkRandom(insertSeed, removeSeed, 1000) { panic("wrong") }
		if !checkRandom(insertSeed, removeSeed, 10000) { panic("wrong") }
		if !checkRandom(insertSeed, removeSeed, 100000) { panic("wrong") }
	}
	fmt.Println("OK")
}

File: jupyter2/problems/recursive_type/tttree/rs/tttree.rs
/** begin my answer */

/** end my answer */

fn add_seq(xs: &[i64], t: TTTree) -> TTTree {
    let mut t = t;
    for x in xs {
        t = tttree_add(*x, t);
    }
    t
}

fn lookup_seq(xs: &[i64], t: &TTTree) -> bool {
    for x in xs {
        if !tttree_lookup(*x, t) {
            return false;
        }
    }
    true
}

fn remove_seq(xs: &[i64], t: TTTree) -> TTTree {
    let mut t = t;
    for x in xs {
        t = tttree_remove(*x, t);
    }
    t
}

fn range(n : i64) -> Vec<i64> {
    (0..n).collect()
}

fn print_vec(header : &str, v : &[i64], n : usize) {
    print!("{}[", header);
    for (i, x) in v.iter().enumerate() {
	if i >= n {
	    print!("...");
	    break;
	}
	if i > 0 {
	    print!(", ");
	}
	print!("{}", x);
    }
    println!("]");
}

/** random sequence */
fn random_seq(n : i64, mut x : i64) -> Vec<i64> {
    let mut s = vec![];
    for _i in 0..n {
        x = (x.wrapping_mul(0x5DEECE66D) + 0xB) & 0xFFFFFFFFFFFF;
        s.push(x >> 17)
    }
    s
}

/** randomly shuffle xs */
fn random_shuffle(seed : i64, xs : Vec<i64>) -> Vec<i64> {
    let ys = random_seq(xs.len() as i64, seed);
    let mut zs : Vec<(i64,i64)> = ys.iter().copied().zip(xs.iter().copied()).collect();
    zs.sort();
    zs.into_iter().map(|(_k, x)| x).collect()
}

/**
make a tree by inserting elements in xs0;
check if all elements in xs1 are in the tree;
then remove all elements in xs1 from the tree
and check if we are left with empty tree */
fn check_seq(xs0: &[i64], xs1: &[i64]) -> bool {
    print_vec("insert in this order: ", xs0, 7);
    print_vec("remove in this order: ", xs1, 7);
    let t = add_seq(xs0, TTTree::Empty);
    let _ = tttree_check(&t);
    if lookup_seq(xs1, &t) {
        match remove_seq(xs1, t) {
	    TTTree::Empty => true,
	    _ => false,
	}
    } else {
        false
    }
}

fn check_random(insert_seed: i64, remove_seed: i64, n: i64) -> bool {
    let rng = range(n);
    let xs0 = random_shuffle(insert_seed, rng.clone());
    let xs1 = random_shuffle(remove_seed, rng);
    check_seq(&xs0, &xs1)
}

fn main() {
    let insert_seed: i64 = 1234;
    let remove_seed: i64 = 123456;
    assert!(check_random(insert_seed, remove_seed, 2));
    assert!(check_random(insert_seed, remove_seed, 3));
    assert!(check_random(insert_seed + 0, remove_seed + 0, 4));
    assert!(check_random(insert_seed + 1, remove_seed + 1, 4));
    assert!(check_random(insert_seed + 2, remove_seed + 2, 4));
    assert!(check_random(insert_seed + 3, remove_seed + 3, 4));
    assert!(check_random(insert_seed + 0, remove_seed + 0, 5));
    assert!(check_random(insert_seed + 1, remove_seed + 1, 5));
    assert!(check_random(insert_seed + 2, remove_seed + 2, 5));
    assert!(check_random(insert_seed + 3, remove_seed + 3, 5));
    assert!(check_random(insert_seed, remove_seed, 10));
    assert!(check_random(insert_seed, remove_seed, 100));
    assert!(check_random(insert_seed, remove_seed, 1000));
    assert!(check_random(insert_seed, remove_seed, 10000));
    assert!(check_random(insert_seed, remove_seed, 100000));
    println!("OK")
}
